{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usefull algorithms.\n",
    "\n",
    "#!/usr/bin/env python\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# This work is licensed under the Creative Commons\n",
    "# Attribution-NonCommercial-ShareAlike 4.0 International License. To view a\n",
    "# copy of this license, visit\n",
    "# http://creativecommons.org/licenses/by-nc-sa/4.0/.\n",
    "# -----------------------------------------------------------------------------\n",
    "import random\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import tabulate\n",
    "from scipy.cluster.hierarchy import to_tree\n",
    "import skbio\n",
    "from skbio import Sequence, DNA, TabularMSA, TreeNode, DistanceMatrix\n",
    "from skbio.alignment import local_pairwise_align_ssw\n",
    "\n",
    "\n",
    "blosum50 = {'A': {'A': 5, 'C': -1, 'D': -2, 'E': -1, 'F': -3, 'G': 0, 'H': -2, 'I': -1, 'K': -1, 'L': -2, 'M': -1, 'N': -1, 'P': -1, 'Q': -1, 'R': -2, 'S': 1, 'T': 0, 'V': 0, 'W': -3, 'Y': -2},\n",
    "'C': {'A': -1, 'C': 13, 'D': -4, 'E': -3, 'F': -2, 'G': -3, 'H': -3, 'I': -2, 'K': -3, 'L': -2, 'M': -2, 'N': -2, 'P': -4, 'Q': -3, 'R': -4, 'S': -1, 'T': -1, 'V': -1, 'W': -5, 'Y': -3},\n",
    "'D': {'A': -2, 'C': -4, 'D': 8, 'E': 2, 'F': -5, 'G': -1, 'H': -1, 'I': -4, 'K': -1, 'L': -4, 'M': -4, 'N': 2, 'P': -1, 'Q': 0, 'R': -2, 'S': 0, 'T': -1, 'V': -4, 'W': -5, 'Y': -3},\n",
    "'E': {'A': -1, 'C': -3, 'D': 2, 'E': 6, 'F': -3, 'G': -3, 'H': 0, 'I': -4, 'K': 1, 'L': -3, 'M': -2, 'N': 0, 'P': -1, 'Q': 2, 'R': 0, 'S': -1, 'T': -1, 'V': -3, 'W': -3, 'Y': -2},\n",
    "'F': {'A': -3, 'C': -2, 'D': -5, 'E': -3, 'F': 8, 'G': -4, 'H': -1, 'I': 0, 'K': -4, 'L': 1, 'M': 0, 'N': -4, 'P': -4, 'Q': -4, 'R': -3, 'S': -3, 'T': -2, 'V': -1, 'W': 1, 'Y': 4},\n",
    "'G': {'A': 0, 'C': -3, 'D': -1, 'E': -3, 'F': -4, 'G': 8, 'H': -2, 'I': -4, 'K': -2, 'L': -4, 'M': -3, 'N': 0, 'P': -2, 'Q': -2, 'R': -3, 'S': 0, 'T': -2, 'V': -4, 'W': -3, 'Y': -3},\n",
    "'H': {'A': -2, 'C': -3, 'D': -1, 'E': 0, 'F': -1, 'G': -2, 'H': 10, 'I': -4, 'K': 0, 'L': -3, 'M': -1, 'N': 1, 'P': -2, 'Q': 1, 'R': 0, 'S': -1, 'T': -2, 'V': -4, 'W': -3, 'Y': 2},\n",
    "'I': {'A': -1, 'C': -2, 'D': -4, 'E': -4, 'F': 0, 'G': -4, 'H': -4, 'I': 5, 'K': -3, 'L': 2, 'M': 2, 'N': -3, 'P': -3, 'Q': -3, 'R': -4, 'S': -3, 'T': -1, 'V': 4, 'W': -3, 'Y': -1},\n",
    "'K': {'A': -1, 'C': -3, 'D': -1, 'E': 1, 'F': -4, 'G': -2, 'H': 0, 'I': -3, 'K': 6, 'L': -3, 'M': -2, 'N': 0, 'P': -1, 'Q': 2, 'R': 3, 'S': 0, 'T': -1, 'V': -3, 'W': -3, 'Y': -2},\n",
    "'L': {'A': -2, 'C': -2, 'D': -4, 'E': -3, 'F': 1, 'G': -4, 'H': -3, 'I': 2, 'K': -3, 'L': 5, 'M': 3, 'N': -4, 'P': -4, 'Q': -2, 'R': -3, 'S': -3, 'T': -1, 'V': 1, 'W': -2, 'Y': -1},\n",
    "'M': {'A': -1, 'C': -2, 'D': -4, 'E': -2, 'F': 0, 'G': -3, 'H': -1, 'I': 2, 'K': -2, 'L': 3, 'M': 7, 'N': -2, 'P': -3, 'Q': 0, 'R': -2, 'S': -2, 'T': -1, 'V': 1, 'W': -1, 'Y': 0},\n",
    "'N': {'A': -1, 'C': -2, 'D': 2, 'E': 0, 'F': -4, 'G': 0, 'H': 1, 'I': -3, 'K': 0, 'L': -4, 'M': -2, 'N': 7, 'P': -2, 'Q': 0, 'R': -1, 'S': 1, 'T': 0, 'V': -3, 'W': -4, 'Y': -2},\n",
    "'P': {'A': -1, 'C': -4, 'D': -1, 'E': -1, 'F': -4, 'G': -2, 'H': -2, 'I': -3, 'K': -1, 'L': -4, 'M': -3, 'N': -2, 'P': 10, 'Q': -1, 'R': -3, 'S': -1, 'T': -1, 'V': -3, 'W': -4, 'Y': -3},\n",
    "'Q': {'A': -1, 'C': -3, 'D': 0, 'E': 2, 'F': -4, 'G': -2, 'H': 1, 'I': -3, 'K': 2, 'L': -2, 'M': 0, 'N': 0, 'P': -1, 'Q': 7, 'R': 1, 'S': 0, 'T': -1, 'V': -3, 'W': -1, 'Y': -1},\n",
    "'R': {'A': -2, 'C': -4, 'D': -2, 'E': 0, 'F': -3, 'G': -3, 'H': 0, 'I': -4, 'K': 3, 'L': -3, 'M': -2, 'N': -1, 'P': -3, 'Q': 1, 'R': 7, 'S': -1, 'T': -1, 'V': -3, 'W': -3, 'Y': -1},\n",
    "'S': {'A': 1, 'C': -1, 'D': 0, 'E': -1, 'F': -3, 'G': 0, 'H': -1, 'I': -3, 'K': 0, 'L': -3, 'M': -2, 'N': 1, 'P': -1, 'Q': 0, 'R': -1, 'S': 5, 'T': 2, 'V': -2, 'W': -4, 'Y': -2},\n",
    "'T': {'A': 0, 'C': -1, 'D': -1, 'E': -1, 'F': -2, 'G': -2, 'H': -2, 'I': -1, 'K': -1, 'L': -1, 'M': -1, 'N': 0, 'P': -1, 'Q': -1, 'R': -1, 'S': 2, 'T': 5, 'V': 0, 'W': -3, 'Y': -2},\n",
    "'V': {'A': 0, 'C': -1, 'D': -4, 'E': -3, 'F': -1, 'G': -4, 'H': -4, 'I': 4, 'K': -3, 'L': 1, 'M': 1, 'N': -3, 'P': -3, 'Q': -3, 'R': -3, 'S': -2, 'T': 0, 'V': 5, 'W': -3, 'Y': -1},\n",
    "'W': {'A': -3, 'C': -5, 'D': -5, 'E': -3, 'F': 1, 'G': -3, 'H': -3, 'I': -3, 'K': -3, 'L': -2, 'M': -1, 'N': -4, 'P': -4, 'Q': -1, 'R': -3, 'S': -4, 'T': -3, 'V': -3, 'W': 15, 'Y': 2},\n",
    "'Y': {'A': -2, 'C': -3, 'D': -3, 'E': -2, 'F': 4, 'G': -3, 'H': 2, 'I': -1, 'K': -2, 'L': -1, 'M': 0, 'N': -2, 'P': -3, 'Q': -1, 'R': -1, 'S': -2, 'T': -2, 'V': -1, 'W': 2, 'Y': 8}}\n",
    "\n",
    "nt_substitution_matrix = {'A': {'A':  1, 'C': -2, 'G': -2, 'T': -2, 'N': 0},\n",
    "                          'C': {'A': -2, 'C':  1, 'G': -2, 'T': -2, 'N': 0},\n",
    "                          'G': {'A': -2, 'C': -2, 'G':  1, 'T': -2, 'N': 0},\n",
    "                          'T': {'A': -2, 'C': -2, 'G': -2, 'T':  1, 'N': 0},\n",
    "                          'N': {'A':  0, 'C':  0, 'G':  0, 'T':  0, 'N': 0 }}\n",
    "\n",
    "traceback_decoding = {1: '\\\\', 2:'|', 3: '-', -1: 'E', 0: '*'}\n",
    "###\n",
    "# pairwise alignment notebook\n",
    "###\n",
    "\n",
    "def hamming_distance(s1, s2):\n",
    "    s1 = Sequence(s1)\n",
    "    s2 = Sequence(s2)\n",
    "    return s1.distance(s2)\n",
    "\n",
    "def show_F(h_sequence, v_sequence, data, hide_zeros=False, nonzero_val=None):\n",
    "    rows = []\n",
    "    col_headers = [c.decode('UTF-8') for c in h_sequence.values]\n",
    "    row_headers = [c.decode('UTF-8') for c in v_sequence.values]\n",
    "    pad_headers = data.shape == (len(row_headers) + 1, len(col_headers) + 1)\n",
    "    if pad_headers:\n",
    "        row_headers = [\" \"] + row_headers\n",
    "        col_headers = [\" \"] + col_headers\n",
    "    for h, d in zip(row_headers, data):\n",
    "        current_row = [\"<b>%s</b>\" % h]\n",
    "        for e in d:\n",
    "            if e == 0:\n",
    "                if hide_zeros:\n",
    "                    current_row.append('')\n",
    "                else:\n",
    "                    current_row.append(0)\n",
    "            else:\n",
    "                if nonzero_val is not None:\n",
    "                    current_row.append(nonzero_val)\n",
    "                else:\n",
    "                    current_row.append(e)\n",
    "        rows.append(current_row)\n",
    "    return tabulate.tabulate(rows, headers=col_headers, tablefmt='html')\n",
    "\n",
    "def show_T(h_sequence, v_sequence, data):\n",
    "    if data.dtype == np.int:\n",
    "        data_ = T = np.full(shape=data.shape, fill_value=\" \", dtype=np.str)\n",
    "        translation_table = {0: \"•\", 1: \"↖\", 2: \"↑\", 3: \"←\"}\n",
    "        for i, row in enumerate(data):\n",
    "            for j, value in enumerate(row):\n",
    "                data_[i, j] =  translation_table[value]\n",
    "    else:\n",
    "        data_ = data\n",
    "    return show_F(h_sequence, v_sequence, data_)\n",
    "\n",
    "def show_substitution_matrix(headers, data):\n",
    "    rows = []\n",
    "    for h, d in zip(headers, data):\n",
    "        current_row = [\"<b>%s</b>\" % h]\n",
    "        for e in d:\n",
    "            current_row.append(e)\n",
    "        rows.append(current_row)\n",
    "    return tabulate.tabulate(rows, headers=headers, tablefmt='html')\n",
    "\n",
    "def format_dynamic_programming_matrix(seq1, seq2, matrix, cell_width=6):\n",
    "    \"\"\" define a function for formatting dynamic programming matrices\n",
    "    \"\"\"\n",
    "    lines = []\n",
    "\n",
    "    if isinstance(seq1, TabularMSA):\n",
    "        seq1 = str(seq1[0])\n",
    "    if isinstance(seq2, TabularMSA):\n",
    "        seq2 = str(seq2[0])\n",
    "    cell_format = \"%\" + str(cell_width) + \"s\"\n",
    "    line_format = cell_format * (len(seq1) + 2)\n",
    "    # print seq1 (start the line with two empty strings)\n",
    "    lines.append(line_format % tuple([' ',' '] + [str(s) for s in list(seq1)]))\n",
    "\n",
    "    # iterate over the rows and print each (starting with the\n",
    "    # corresponding base in sequence2)\n",
    "    for row, base in zip(matrix,' ' + seq2):\n",
    "        row_list = [base]\n",
    "        for s in row:\n",
    "            if isinstance(s, np.float):\n",
    "                s = str(s)\n",
    "            else:\n",
    "                s = s.decode('ascii')\n",
    "            row_list.append(s)\n",
    "        line = line_format % tuple(row_list)\n",
    "        lines.append(line)\n",
    "\n",
    "    return '\\n'.join(lines)\n",
    "\n",
    "def format_traceback_matrix(seq1, seq2, matrix, cell_width=6):\n",
    "    if isinstance(seq1, TabularMSA):\n",
    "        seq1 = str(seq1[0])\n",
    "    if isinstance(seq2, TabularMSA):\n",
    "        seq2 = str(seq2[0])\n",
    "    translated_m = np.chararray(matrix.shape)\n",
    "    for i in range(matrix.shape[0]):\n",
    "        for j in range(matrix.shape[1]):\n",
    "            translated_m[i, j] = traceback_decoding[matrix[i, j]]\n",
    "\n",
    "    return format_dynamic_programming_matrix(seq1, seq2, translated_m,\n",
    "                                             cell_width)\n",
    "\n",
    "\n",
    "def format_dynamic_programming_matrix_subset(seq1,seq2,matrix, cell_width=6, num_positions=10):\n",
    "    \"\"\" return first num_positions x num_positions of dynamic programming matrix\n",
    "    \"\"\"\n",
    "    lines = []\n",
    "    if isinstance(seq1, TabularMSA):\n",
    "        seq1 = str(seq1[0])\n",
    "    if isinstance(seq2, TabularMSA):\n",
    "        seq2 = str(seq2[0])\n",
    "    cell_format = \"%\" + str(cell_width) + \"s\"\n",
    "    line_format = cell_format * (len(seq1[:num_positions]) + 1)\n",
    "    # print seq1 (start the line with two empty strings)\n",
    "    lines.append(line_format % tuple([' '] + map(str,list(seq1[:num_positions]))))\n",
    "\n",
    "    # iterate over the rows and print each (starting with the\n",
    "    # corresponding base in sequence2)\n",
    "    for row, base in zip(matrix[:num_positions],' ' + seq2[:num_positions]):\n",
    "        lines.append(line_format % tuple([base] + map(str,row[:num_positions])))\n",
    "\n",
    "    return '\\n'.join(lines)\n",
    "\n",
    "\n",
    "def generate_score_matrix(seq1,seq2,substitution_matrix):\n",
    "    # Initialize a matrix to use for storing the scores\n",
    "    score_matrix = []\n",
    "    # Iterate over the amino acids in sequence two (which will correspond\n",
    "    # to the vertical sequence in the matrix)\n",
    "    for aa2 in seq2:\n",
    "        # Initialize the current row of the matrix\n",
    "        current_row = []\n",
    "        # Iterate over the amino acids in sequence one (which will\n",
    "        # correspond to the horizontal sequence in the matrix)\n",
    "        for aa1 in seq1:\n",
    "            # score as 1 if the bases are equal and 0 if they're not\n",
    "            current_row.append(substitution_matrix[aa1][aa2])\n",
    "        # append the current row to the matrix\n",
    "        score_matrix.append(current_row)\n",
    "    return score_matrix\n",
    "\n",
    "##\n",
    "# Needleman-Wunsch alignment\n",
    "##\n",
    "\n",
    "def generate_nw_and_traceback_matrices(seq1,seq2,gap_penalty,substitution_matrix):\n",
    "\n",
    "    # Initialize a matrix to use for scoring the alignment and for tracing\n",
    "    # back the best alignment\n",
    "    nw_matrix = [[-gap_penalty * i for i in range(0,len(seq1)+1)]]\n",
    "    traceback_matrix = [[None] + ['-' for i in range(0,len(seq1))]]\n",
    "    # Iterate over the amino acids in sequence two (which will correspond\n",
    "    # to the vertical sequence in the matrix)\n",
    "    # Note that i corresponds to column numbers, as in the 'Biological Sequence\n",
    "    # Analysis' example\n",
    "    for i,aa2 in zip(range(1,len(seq2)+1),seq2):\n",
    "        # Initialize the current row of the matrix\n",
    "        current_row = [i * -gap_penalty]\n",
    "        current_traceback_matrix_row = ['|']\n",
    "        # Iterate over the amino acids in sequence one (which will\n",
    "        # correspond to the horizontal sequence in the matrix)\n",
    "        # Note that j corresponds to row numbers, as in the 'Biological Sequence\n",
    "        # Analysis' example from class\n",
    "        for j,aa1 in zip(range(1,len(seq1)+1),seq1):\n",
    "            substitution_score = substitution_matrix[aa1][aa2]\n",
    "            diag_score = (nw_matrix[i-1][j-1] + substitution_score,'\\\\')\n",
    "            up_score = (nw_matrix[i-1][j] - gap_penalty,'|')\n",
    "            left_score = (current_row[-1] - gap_penalty,'-')\n",
    "            best_score = max_score_tuple(diag_score, up_score, left_score)\n",
    "            current_row.append(best_score[0])\n",
    "            current_traceback_matrix_row.append(best_score[1])\n",
    "        # append the current row to the matrix\n",
    "        nw_matrix.append(current_row)\n",
    "        traceback_matrix.append(current_traceback_matrix_row)\n",
    "    return nw_matrix, traceback_matrix\n",
    "\n",
    "\n",
    "def nw_traceback(traceback_matrix,nw_matrix,seq1,seq2,gap_character='-'):\n",
    "\n",
    "    aligned_seq1 = []\n",
    "    aligned_seq2 = []\n",
    "\n",
    "    current_row = len(traceback_matrix) - 1\n",
    "    current_col = len(traceback_matrix[0]) - 1\n",
    "\n",
    "    best_score = nw_matrix[current_row][current_col]\n",
    "\n",
    "    while True:\n",
    "        current_value = traceback_matrix[current_row][current_col]\n",
    "\n",
    "        if current_value == '\\\\':\n",
    "            aligned_seq1.append(seq1[current_col-1])\n",
    "            aligned_seq2.append(seq2[current_row-1])\n",
    "            current_row -= 1\n",
    "            current_col -= 1\n",
    "        elif current_value == '|':\n",
    "            aligned_seq1.append('-')\n",
    "            aligned_seq2.append(seq2[current_row-1])\n",
    "            current_row -= 1\n",
    "        elif current_value == '-':\n",
    "            aligned_seq1.append(seq1[current_col-1])\n",
    "            aligned_seq2.append('-')\n",
    "            current_col -= 1\n",
    "        elif current_value == None:\n",
    "            break\n",
    "        else:\n",
    "            raise ValueError(\"Invalid value in traceback matrix: %s\" % current_value)\n",
    "\n",
    "    return ''.join(aligned_seq1[::-1]), ''.join(aligned_seq2[::-1]), best_score\n",
    "\n",
    "\n",
    "def nw_align(seq1, seq2, gap_penalty, substitution_matrix):\n",
    "    \"\"\" Perform Needleman-Wunsch alignment of seq1 and seq2\n",
    "    \"\"\"\n",
    "    nw_matrix, traceback_matrix = generate_nw_and_traceback_matrices(\n",
    "                                    seq1, seq2, gap_penalty, substitution_matrix)\n",
    "    aligned_seq1, aligned_seq2, score = nw_traceback(traceback_matrix,nw_matrix,seq1,seq2)\n",
    "    return aligned_seq1, aligned_seq2, score\n",
    "\n",
    "def nw_align_nt(seq1, seq2, gap_penalty=8, substitution_matrix=nt_substitution_matrix):\n",
    "    \"\"\"Globally align two nucleotide seqs (Needleman-Wunsch w single gap scoring)\n",
    "       Parameters\n",
    "       ----------\n",
    "       sequence1 : string\n",
    "           The first unaligned sequence\n",
    "       sequence2 : string\n",
    "           The second unaligned sequence\n",
    "       gap_penalty : int, float, optional\n",
    "           penalty for inserting a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       substitution_matrix: 2D dict (or similar), optional\n",
    "           lookup for substitution scores (these values are added to the\n",
    "           previous best alignment score)\n",
    "       Returns\n",
    "       -------\n",
    "       string\n",
    "          The first aligned sequence\n",
    "       string\n",
    "          The second aligned sequence\n",
    "       float\n",
    "          The score of the alignment\n",
    "    \"\"\"\n",
    "    return nw_align(seq1, seq2, gap_penalty, substitution_matrix)\n",
    "\n",
    "\n",
    "##\n",
    "# Smith-Waterman alignment\n",
    "##\n",
    "\n",
    "\n",
    "def generate_sw_and_traceback_matrices(seq1,seq2,gap_penalty,substitution_matrix):\n",
    "    # Initialize a matrix to use for scoring the alignment and for tracing\n",
    "    # back the best alignment\n",
    "    sw_matrix = [[0 for i in range(0,len(seq1)+1)]]\n",
    "    traceback_matrix = [[None] + [None for i in range(0,len(seq1))]]\n",
    "    # Iterate over the amino acids in sequence two (which will correspond\n",
    "    # to the vertical sequence in the matrix)\n",
    "    # Note that i corresponds to column numbers, as in the 'Biological Sequence\n",
    "    # Analysis' example from class\n",
    "    for i,aa2 in zip(range(1,len(seq2)+1),seq2):\n",
    "        # Initialize the current row of the matrix\n",
    "        current_row = [0]\n",
    "        current_traceback_matrix_row = [None]\n",
    "        # Iterate over the amino acids in sequence one (which will\n",
    "        # correspond to the horizontal sequence in the matrix)\n",
    "        # Note that j corresponds to row numbers, as in the 'Biological Sequence\n",
    "        # Analysis' example from class\n",
    "        new_alignment_score = (0,None)\n",
    "        for j,aa1 in zip(range(1,len(seq1)+1),seq1):\n",
    "            substitution_score = substitution_matrix[aa1][aa2]\n",
    "            diag_score = (sw_matrix[i-1][j-1] + substitution_score,'\\\\')\n",
    "            up_score = (sw_matrix[i-1][j] - gap_penalty,'|')\n",
    "            left_score = (current_row[-1] - gap_penalty,'-')\n",
    "            best_score = max_score_tuple(diag_score, up_score, left_score,\n",
    "                                         new_alignment_score)\n",
    "            current_row.append(best_score[0])\n",
    "            current_traceback_matrix_row.append(best_score[1])\n",
    "        # append the current row to the matrix\n",
    "        sw_matrix.append(current_row)\n",
    "        traceback_matrix.append(current_traceback_matrix_row)\n",
    "    return sw_matrix, traceback_matrix\n",
    "\n",
    "\n",
    "def sw_traceback(traceback_matrix,sw_matrix,seq1,seq2,gap_character='-'):\n",
    "\n",
    "    aligned_seq1 = []\n",
    "    aligned_seq2 = []\n",
    "\n",
    "    current_row = None\n",
    "    current_col = None\n",
    "    best_score = 0\n",
    "    for i in range(len(sw_matrix[0])):\n",
    "        for j in range(len(sw_matrix)):\n",
    "            current_score = sw_matrix[j][i]\n",
    "            if current_score > best_score:\n",
    "                best_score = current_score\n",
    "                current_row = j\n",
    "                current_col = i\n",
    "\n",
    "    while True:\n",
    "        current_value = traceback_matrix[current_row][current_col]\n",
    "\n",
    "        if current_value == '\\\\':\n",
    "            aligned_seq1.append(seq1[current_col-1])\n",
    "            aligned_seq2.append(seq2[current_row-1])\n",
    "            current_row -= 1\n",
    "            current_col -= 1\n",
    "        elif current_value == '|':\n",
    "            aligned_seq1.append('-')\n",
    "            aligned_seq2.append(seq2[current_row-1])\n",
    "            current_row -= 1\n",
    "        elif current_value == '-':\n",
    "            aligned_seq1.append(seq1[current_col-1])\n",
    "            aligned_seq2.append('-')\n",
    "            current_col -= 1\n",
    "        elif current_value == None:\n",
    "            break\n",
    "        else:\n",
    "            raise ValueError(\"Invalid value in traceback matrix: %s\" % current_value)\n",
    "\n",
    "    return ''.join(aligned_seq1[::-1]), ''.join(aligned_seq2[::-1]), best_score, current_col, current_row\n",
    "\n",
    "def sw_multiple_traceback(traceback_matrix,sw_matrix,seq1,seq2,gap_character='-'):\n",
    "\n",
    "    aligned_seq1 = []\n",
    "    aligned_seq2 = []\n",
    "\n",
    "    current_row = None\n",
    "    current_col = None\n",
    "    best_scores = []\n",
    "    for i in range(len(sw_matrix[0])):\n",
    "        for j in range(len(sw_matrix)):\n",
    "            try:\n",
    "                new_alignment = traceback_matrix[j+1][i+1] == None\n",
    "            except IndexError:\n",
    "                new_alignment = True\n",
    "            if new_alignment:\n",
    "                current_score = sw_matrix[j][i]\n",
    "                current_row = j\n",
    "                current_col = i\n",
    "                best_scores.append((current_score, current_row, current_col))\n",
    "    best_scores.sort()\n",
    "    best_scores.reverse()\n",
    "\n",
    "    results = []\n",
    "    for current_score, current_row, current_col in best_scores[:10]:\n",
    "        while True:\n",
    "            current_value = traceback_matrix[current_row][current_col]\n",
    "\n",
    "            if current_value == '\\\\':\n",
    "                aligned_seq1.append(seq1[current_col-1])\n",
    "                aligned_seq2.append(seq2[current_row-1])\n",
    "                current_row -= 1\n",
    "                current_col -= 1\n",
    "            elif current_value == '|':\n",
    "                aligned_seq1.append('-')\n",
    "                aligned_seq2.append(seq2[current_row-1])\n",
    "                current_row -= 1\n",
    "            elif current_value == '-':\n",
    "                aligned_seq1.append(seq1[current_col-1])\n",
    "                aligned_seq2.append('-')\n",
    "                current_col -= 1\n",
    "            elif current_value == None:\n",
    "                break\n",
    "            else:\n",
    "                raise ValueError(\"Invalid value in traceback matrix: %s\" % current_value)\n",
    "        results.append((''.join(aligned_seq1[::-1]), ''.join(aligned_seq2[::-1]), current_score, current_col, current_row))\n",
    "\n",
    "    return results\n",
    "\n",
    "\n",
    "def sw_align(sequence1, sequence2, gap_penalty, substitution_matrix):\n",
    "    sw_matrix, traceback_matrix = generate_sw_and_traceback_matrices(sequence1,\n",
    "                                                                 sequence2,\n",
    "                                                                 gap_penalty,\n",
    "                                                                 substitution_matrix)\n",
    "\n",
    "    return sw_traceback(traceback_matrix,sw_matrix,sequence1,sequence2)\n",
    "\n",
    "\n",
    "def generate_sw_and_traceback_matrices_affine_gap(seq1, seq2, gap_open_penalty, gap_extend_penalty, substitution_matrix):\n",
    "    # Initialize a matrix to use for scoring the alignment and for tracing\n",
    "    # back the best alignment\n",
    "    sw_matrix = [[0 for i in range(0,len(seq1)+1)]]\n",
    "    traceback_matrix = [[None] + [None for i in range(0,len(seq1))]]\n",
    "    # Iterate over the amino acids in sequence two (which will correspond\n",
    "    # to the vertical sequence in the matrix)\n",
    "    # Note that i corresponds to column numbers, as in the 'Biological Sequence\n",
    "    # Analysis' example from class\n",
    "    for i,aa2 in zip(range(1,len(seq2)+1),seq2):\n",
    "        # Initialize the current row of the matrix\n",
    "        current_row = [0]\n",
    "        current_traceback_matrix_row = [None]\n",
    "        # Iterate over the amino acids in sequence one (which will\n",
    "        # correspond to the horizontal sequence in the matrix)\n",
    "        # Note that j corresponds to row numbers, as in the 'Biological Sequence\n",
    "        # Analysis' example from class\n",
    "        new_alignment_score = (0,None)\n",
    "        for j,aa1 in zip(range(1,len(seq1)+1),seq1):\n",
    "            substitution_score = substitution_matrix[aa1][aa2]\n",
    "            diag_score = (sw_matrix[i-1][j-1] + substitution_score,'\\\\')\n",
    "            if traceback_matrix[i-1][j] == '|':\n",
    "                # gap extend, because the cell above was also a gap\n",
    "                up_score = (sw_matrix[i-1][j] - gap_extend_penalty,'|')\n",
    "            else:\n",
    "                # gap open, because the cell above was not a gap\n",
    "                up_score = (sw_matrix[i-1][j] - gap_open_penalty,'|')\n",
    "            if current_traceback_matrix_row[-1] == '-':\n",
    "                # gap extend, because the cell to the left was also a gap\n",
    "                left_score = (current_row[-1] - gap_extend_penalty,'-')\n",
    "            else:\n",
    "                # gap open, because the cell to the left was not a gap\n",
    "                left_score = (current_row[-1] - gap_open_penalty,'-')\n",
    "            best_score = max_score_tuple(diag_score, up_score, left_score,\n",
    "                                         new_alignment_score)\n",
    "            current_row.append(best_score[0])\n",
    "            current_traceback_matrix_row.append(best_score[1])\n",
    "        # append the current row to the matrix\n",
    "        sw_matrix.append(current_row)\n",
    "        traceback_matrix.append(current_traceback_matrix_row)\n",
    "    return sw_matrix, traceback_matrix\n",
    "\n",
    "def max_score_tuple(*scores):\n",
    "    return max(scores, key=lambda x: x[0])\n",
    "\n",
    "def sw_align_affine_gap(sequence1, sequence2, gap_open_penalty, gap_extend_penalty, substitution_matrix):\n",
    "    sw_matrix, traceback_matrix = generate_sw_and_traceback_matrices_affine_gap(sequence1,\n",
    "                                                                 sequence2,\n",
    "                                                                 gap_open_penalty,\n",
    "                                                                 gap_extend_penalty,\n",
    "                                                                 substitution_matrix)\n",
    "\n",
    "    return sw_traceback(traceback_matrix,sw_matrix,sequence1,sequence2)\n",
    "\n",
    "def sw_align_affine_gap_nt(sequence1, sequence2, gap_open_penalty=5,\n",
    "                        gap_extend_penalty=2, substitution_matrix=None):\n",
    "    \"\"\"Locally align two nucleotide seqs (Smith-Waterman w affine gap scoring)\n",
    "       Parameters\n",
    "       ----------\n",
    "       sequence1 : string\n",
    "           The first unaligned sequence\n",
    "       sequence2 : string\n",
    "           The second unaligned sequence\n",
    "       gap_open_penalty : int, float, optional\n",
    "           penalty for opening a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       gap_extend_penalty : int, float, optional\n",
    "           penalty for extending a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       substitution_matrix: 2D dict (or similar), optional\n",
    "           lookup for substitution scores (these values are added to the\n",
    "           previous best alignment score); default is nt_substitution_matrix\n",
    "       Returns\n",
    "       -------\n",
    "       string\n",
    "          The first aligned sequence\n",
    "       string\n",
    "          The second aligned sequence\n",
    "       float\n",
    "          The score of the alignment\n",
    "       int\n",
    "          The start position of the alignment in sequence 1\n",
    "       int\n",
    "          The start position of the alignment in sequence 2\n",
    "       Examples\n",
    "       --------\n",
    "       >>> from iab.algorithms import sw_align_affine_gap_nt\n",
    "       >>> s1 = \"GCGTGCCTAAGGTATGCAAG\"\n",
    "       >>> s2 = \"ACGTGCCTAGGTACGCAAG\"\n",
    "       >>> a1, a2, score, a1_start, a2_start = sw_align_affine_gap_nt(s1, s2)\n",
    "       >>> print(a1)\n",
    "       CGTGCCTAAGGTATGCAAG\n",
    "       >>> print(a2)\n",
    "       CGTGCCT-AGGTACGCAAG\n",
    "    \"\"\"\n",
    "    if substitution_matrix is None:\n",
    "        substitution_matrix = nt_substitution_matrix\n",
    "\n",
    "    return sw_align_affine_gap(sequence1, sequence2, gap_open_penalty,\n",
    "                               gap_extend_penalty, substitution_matrix)\n",
    "\n",
    "def sw_multiple_align_affine_gap_nt(sequence1, sequence2, gap_open_penalty=5,\n",
    "                        gap_extend_penalty=2,\n",
    "                        substitution_matrix=nt_substitution_matrix):\n",
    "    sw_matrix, traceback_matrix = generate_sw_and_traceback_matrices_affine_gap(sequence1,\n",
    "                                                                 sequence2,\n",
    "                                                                 gap_open_penalty,\n",
    "                                                                 gap_extend_penalty,\n",
    "                                                                 substitution_matrix)\n",
    "\n",
    "    return sw_multiple_traceback(traceback_matrix,sw_matrix,sequence1,sequence2)\n",
    "\n",
    "def sw_align_affine_gap_pr(sequence1, sequence2, gap_open_penalty=11,\n",
    "                        gap_extend_penalty=1,\n",
    "                        substitution_matrix=blosum50):\n",
    "    \"\"\"Locally align two protein seqs (Smith-Waterman w affine gap scoring)\n",
    "       Parameters\n",
    "       ----------\n",
    "       sequence1 : string\n",
    "           The first unaligned sequence\n",
    "       sequence2 : string\n",
    "           The second unaligned sequence\n",
    "       gap_open_penalty : int, float, optional\n",
    "           penalty for opening a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       gap_extend_penalty : int, float, optional\n",
    "           penalty for extending a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       substitution_matrix: 2D dict (or similar), optional\n",
    "           lookup for substitution scores (these values are added to the\n",
    "           previous best alignment score)\n",
    "       Returns\n",
    "       -------\n",
    "       string\n",
    "          The first aligned sequence\n",
    "       string\n",
    "          The second aligned sequence\n",
    "       float\n",
    "          The score of the alignment\n",
    "       int\n",
    "          The start position of the alignment in sequence 1\n",
    "       int\n",
    "          The start position of the alignment in sequence 2\n",
    "    \"\"\"\n",
    "    return sw_align_affine_gap(sequence1, sequence2, gap_open_penalty,\n",
    "                               gap_extend_penalty, substitution_matrix)\n",
    "\n",
    "def sw_align_nt(seq1, seq2, gap_penalty=8, substitution_matrix=nt_substitution_matrix):\n",
    "    \"\"\"Locally align two nucleotide seqs (Smith-Waterman w single gap scoring)\n",
    "       Parameters\n",
    "       ----------\n",
    "       sequence1 : string\n",
    "           The first unaligned sequence\n",
    "       sequence2 :\n",
    "           The second unaligned sequence\n",
    "       gap_penalty : int, float, optional\n",
    "           penalty for inserting a gap (this is substracted from previous best\n",
    "           alignment score, so is typically positive)\n",
    "       substitution_matrix: 2D dict (or similar), optional\n",
    "           lookup for substitution scores (these values are added to the\n",
    "           previous best alignment score)\n",
    "       Returns\n",
    "       -------\n",
    "       string\n",
    "          The first aligned sequence\n",
    "       string\n",
    "          The second aligned sequence\n",
    "       float\n",
    "          The score of the alignment\n",
    "       int\n",
    "          The start position of the alignment in sequence 1\n",
    "       int\n",
    "          The start position of the alignment in sequence 2\n",
    "    \"\"\"\n",
    "    return sw_align(seq1, seq2, gap_penalty, substitution_matrix)\n",
    "\n",
    "def align(sequence1, sequence2, gap_penalty, substitution_matrix, local):\n",
    "    if local:\n",
    "        return sw_align(sequence1, sequence2, gap_penalty, substitution_matrix)\n",
    "    else:\n",
    "        return nw_align(sequence1, sequence2, gap_penalty, substitution_matrix)\n",
    "\n",
    "##\n",
    "# DB searching notebook\n",
    "##\n",
    "\n",
    "def local_alignment_search(queries, reference_db, n=5,\n",
    "                           aligner=local_pairwise_align_ssw):\n",
    "    results = []\n",
    "    indices = []\n",
    "    for q in queries:\n",
    "        # first we'll compute all of the alignments and their associated scores\n",
    "        hits = []\n",
    "        for r in reference_db:\n",
    "            aln, score, _ = aligner(q, r)\n",
    "            hits.append([r.metadata['id'], score, aln,\n",
    "                         r.metadata['taxonomy']])\n",
    "        # then we reverse-sort them by score, and return the n highest\n",
    "        # scoring alignments (this needs to be updated so we only\n",
    "        # ever keep track of the n highest scoring alignments)\n",
    "        best_hits = sorted(hits, key=lambda e: e[1], reverse=True)[:n]\n",
    "        if len(best_hits) == 0:\n",
    "            # if there are no hits, log that information\n",
    "            indices.append((q.metadata['id'], \"no matches\"))\n",
    "            results.append((\"n/a\", np.nan, np.nan, np.nan))\n",
    "        else:\n",
    "            # otherwise compile and track some information about the n\n",
    "            # best hits\n",
    "            for r_id, score, aln, r_tax in best_hits:\n",
    "                percent_similarity = (100 * (1. - aln[0].distance(aln[1])))\n",
    "                aln_length = aln.shape[1]\n",
    "                indices.append((q.metadata['id'], r_id))\n",
    "                results.append((r_tax, percent_similarity,\n",
    "                                aln_length, score))\n",
    "    index = pd.MultiIndex.from_tuples(indices, names=['query', 'reference'])\n",
    "    columns = ['reference taxonomy', 'percent similarity',\n",
    "               'alignment length', 'score']\n",
    "    results = pd.DataFrame(results, index=index, columns=columns)\n",
    "    return results\n",
    "\n",
    "def heuristic_local_alignment_search_random(\n",
    "        queries, reference_db, p, n=5, aligner=local_pairwise_align_ssw):\n",
    "    k = int(p * len(reference_db))\n",
    "    database_subset = random.sample(reference_db, k)\n",
    "    return local_alignment_search(queries, database_subset, n=n, aligner=aligner)\n",
    "\n",
    "def heuristic_local_alignment_search_gc(\n",
    "        queries, reference_db, database_subset_size, n=5,\n",
    "        reference_db_gc_contents=None,\n",
    "        aligner=local_pairwise_align_ssw):\n",
    "    results = []\n",
    "    if reference_db_gc_contents is None:\n",
    "        reference_db_gc_contents = \\\n",
    "         {r.metadata['id'] : r.gc_content() for r in reference_db}\n",
    "    for q in queries:\n",
    "        query_gc_content = q.gc_content()\n",
    "        database_subset = []\n",
    "        for r in reference_db:\n",
    "            ref_gc_content = reference_db_gc_contents[r.metadata['id']]\n",
    "            # find the difference in GC content between the reference and\n",
    "            # query. we'll sort and select our reference sequences by this\n",
    "            # value\n",
    "            database_subset.append((abs(ref_gc_content - query_gc_content), r))\n",
    "        database_subset.sort(key=lambda x: x[0])\n",
    "        database_subset = [e[1] for e in database_subset[:database_subset_size]]\n",
    "        results.append(local_alignment_search(\n",
    "            [q], database_subset, n=n, aligner=aligner))\n",
    "    return pd.concat(results)\n",
    "\n",
    "def shuffle_sequence(sequence):\n",
    "    # generate a list of the position indices (numbers) in sequence\n",
    "    randomized_order = list(range(len(sequence)))\n",
    "    # randomly rearrange the order of that list\n",
    "    random.shuffle(randomized_order)\n",
    "    # return a new sequence, where the positions are shuffled\n",
    "    return sequence[randomized_order]\n",
    "\n",
    "def generate_random_score_distribution(sequence1,\n",
    "                                       sequence2,\n",
    "                                       n=99,\n",
    "                                       aligner=local_pairwise_align_ssw):\n",
    "    scores = []\n",
    "    # iterate n times\n",
    "    for i in range(n):\n",
    "        # generate a randomized version of the first sequence\n",
    "        random_sequence = shuffle_sequence(sequence1)\n",
    "        # align that randomized sequence against the second sequence\n",
    "        # and save its score\n",
    "        _, score, _ = aligner(random_sequence, sequence2)\n",
    "        scores.append(score)\n",
    "    # return the n randomized alignment scores\n",
    "    return scores\n",
    "\n",
    "def fraction_better_or_equivalent_alignments(sequence1,\n",
    "                                             sequence2,\n",
    "                                             n = 99,\n",
    "                                             aligner=local_pairwise_align_ssw):\n",
    "    # align sequence1 and sequence2 and store the score of the alignment\n",
    "    _, actual_score, _ = aligner(sequence1, sequence2)\n",
    "    # compute the distribution of randomized scores\n",
    "    random_scores = generate_random_score_distribution(sequence1,\n",
    "                                                       sequence2,\n",
    "                                                       n,\n",
    "                                                       aligner=aligner)\n",
    "\n",
    "    # count the number of random scores that are at least as good as our\n",
    "    # actual score\n",
    "    count_better = 0\n",
    "    for s in random_scores:\n",
    "        if s >= actual_score:\n",
    "            count_better += 1\n",
    "    # return the number of times we observe a score at least as good as the\n",
    "    # random score divided by the number of scores we computed. we add one\n",
    "    # to the numerator and denominator to account for our actual_score\n",
    "    return (count_better + 1) / (n + 1)\n",
    "\n",
    "\n",
    "def fraction_shared_kmers(kmer_freqs1, kmer_freqs2):\n",
    "    \"\"\"Compute the fraction of kmers in kmer_freqs1 that are also in kmer_freqs2\n",
    "    Parameters\n",
    "    ----------\n",
    "    kmer_freqs1, kmer_freqs2\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "    Raises\n",
    "    ------\n",
    "    ValueError\n",
    "        If k < 1.\n",
    "    Notes\n",
    "    -----\n",
    "    k-mer counts are not incorporated in this distance metric.\n",
    "    \"\"\"\n",
    "    sequence1_kmers = set(kmer_freqs1)\n",
    "    num_sequence1_kmers = len(sequence1_kmers)\n",
    "    sequence2_kmers = set(kmer_freqs2)\n",
    "    shared_kmers = sequence1_kmers & sequence2_kmers\n",
    "    return len(shared_kmers) / num_sequence1_kmers\n",
    "\n",
    "def heuristic_local_alignment_search_kmers(\n",
    "        queries, reference_db, database_subset_size, k, n=5,\n",
    "        reference_db_kmer_frequencies=None,\n",
    "        aligner=local_pairwise_align_ssw):\n",
    "    results = []\n",
    "    if reference_db_kmer_frequencies is None:\n",
    "        reference_db_kmer_frequencies = \\\n",
    "         {r.metadata['id'] : r.kmer_frequencies(k=k, overlap=True) for r in reference_db}\n",
    "    for q in queries:\n",
    "        query_kmer_frequency = q.kmer_frequencies(k=k, overlap=True)\n",
    "        database_subset = []\n",
    "        for r in reference_db:\n",
    "            ref_kmer_frequency = reference_db_kmer_frequencies[r.metadata['id']]\n",
    "            s = fraction_shared_kmers(query_kmer_frequency, ref_kmer_frequency)\n",
    "            database_subset.append((s, r))\n",
    "        database_subset.sort(key=lambda x: x[0], reverse=True)\n",
    "        database_subset = [e[1] for e in database_subset[:database_subset_size]]\n",
    "        results.append(local_alignment_search(\n",
    "            [q], database_subset, n=n, aligner=aligner))\n",
    "    return pd.concat(results)\n",
    "\n",
    "\n",
    "## MSA notebook\n",
    "\n",
    "def kmer_distance(sequence1, sequence2, k=3, overlap=True):\n",
    "    \"\"\"Compute the kmer distance between a pair of sequences\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequence1 : skbio.Sequence\n",
    "    sequence2 : skbio.Sequence\n",
    "    k : int, optional\n",
    "        The word length.\n",
    "    overlapping : bool, optional\n",
    "        Defines whether the k-words should be overlapping or not\n",
    "        overlapping.\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        Fraction of the set of k-mers from both sequence1 and\n",
    "        sequence2 that are unique to either sequence1 or\n",
    "        sequence2.\n",
    "    Raises\n",
    "    ------\n",
    "    ValueError\n",
    "        If k < 1.\n",
    "    Notes\n",
    "    -----\n",
    "    k-mer counts are not incorporated in this distance metric.\n",
    "    \"\"\"\n",
    "    sequence1_kmers = set(map(str, sequence1.iter_kmers(k, overlap)))\n",
    "    sequence2_kmers = set(map(str, sequence2.iter_kmers(k, overlap)))\n",
    "    all_kmers = sequence1_kmers | sequence2_kmers\n",
    "    shared_kmers = sequence1_kmers & sequence2_kmers\n",
    "    number_unique = len(all_kmers) - len(shared_kmers)\n",
    "    fraction_unique = number_unique / len(all_kmers)\n",
    "    return fraction_unique\n",
    "\n",
    "def guide_tree_from_sequences(sequences,\n",
    "                              metric=kmer_distance,\n",
    "                              display_tree = False):\n",
    "    \"\"\" Build a UPGMA tree by applying metric to sequences\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : list of skbio.Sequence objects (or subclasses)\n",
    "      The sequences to be represented in the resulting guide tree.\n",
    "    metric : function\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects.\n",
    "    display_tree : bool, optional\n",
    "      Print the tree before returning.\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.TreeNode\n",
    "    \"\"\"\n",
    "    guide_dm = DistanceMatrix.from_iterable(\n",
    "                    sequences, metric=metric, key='id')\n",
    "    guide_lm = sp.cluster.hierarchy.average(guide_dm.condensed_form())\n",
    "    guide_tree = to_tree(guide_lm)\n",
    "    if display_tree:\n",
    "        guide_d = sp.cluster.hierarchy.dendrogram(guide_lm, labels=guide_dm.ids, orientation='right',\n",
    "               link_color_func=lambda x: 'black')\n",
    "    return guide_tree\n",
    "\n",
    "def progressive_msa(sequences, pairwise_aligner, guide_tree=None,\n",
    "                    metric=kmer_distance):\n",
    "    \"\"\" Perform progressive msa of sequences\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : skbio.SequenceCollection\n",
    "        The sequences to be aligned.\n",
    "    metric : function, optional\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects. This will be used to build a guide tree if one\n",
    "      is not provided.\n",
    "    guide_tree : skbio.TreeNode, optional\n",
    "        The tree that should be used to guide the alignment process.\n",
    "    pairwise_aligner : function\n",
    "        Function that should be used to perform the pairwise alignments,\n",
    "        for example skbio.alignment.global_pairwise_align_nucleotide. Must\n",
    "        support skbio.Sequence objects or skbio.TabularMSA objects\n",
    "        as input.\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.TabularMSA\n",
    "    \"\"\"\n",
    "\n",
    "    if guide_tree is None:\n",
    "        guide_dm = DistanceMatrix.from_iterable(\n",
    "                        sequences, metric=metric, key='id')\n",
    "        guide_lm = sp.cluster.hierarchy.average(guide_dm.condensed_form())\n",
    "        guide_tree = TreeNode.from_linkage_matrix(guide_lm, guide_dm.ids)\n",
    "\n",
    "    seq_lookup = {s.metadata['id']: s for i, s in enumerate(sequences)}\n",
    "    c1, c2 = guide_tree.children\n",
    "    if c1.is_tip():\n",
    "        c1_aln = seq_lookup[c1.name]\n",
    "    else:\n",
    "        c1_aln = progressive_msa(sequences, pairwise_aligner, c1)\n",
    "\n",
    "    if c2.is_tip():\n",
    "        c2_aln = seq_lookup[c2.name]\n",
    "    else:\n",
    "        c2_aln = progressive_msa(sequences, pairwise_aligner, c2)\n",
    "\n",
    "    alignment, _, _ = pairwise_aligner(c1_aln, c2_aln)\n",
    "    # this is a temporary hack as the aligners in skbio 0.4.1 are dropping\n",
    "    # metadata - this makes sure that the right metadata is associated with\n",
    "    # the sequence after alignment\n",
    "    if isinstance(c1_aln, Sequence):\n",
    "        alignment[0].metadata = c1_aln.metadata\n",
    "        len_c1_aln = 1\n",
    "    else:\n",
    "        for i in range(len(c1_aln)):\n",
    "            alignment[i].metadata = c1_aln[i].metadata\n",
    "        len_c1_aln = len(c1_aln)\n",
    "    if isinstance(c2_aln, Sequence):\n",
    "        alignment[1].metadata = c2_aln.metadata\n",
    "    else:\n",
    "        for i in range(len(c2_aln)):\n",
    "            alignment[len_c1_aln + i].metadata = c2_aln[i].metadata\n",
    "\n",
    "    return alignment\n",
    "\n",
    "def tree_from_distance_matrix(dm, metric):\n",
    "    if metric == \"upgma\":\n",
    "        lm = sp.cluster.hierarchy.average(dm.condensed_form())\n",
    "        tree = TreeNode.from_linkage_matrix(lm, dm.ids)\n",
    "    elif metric == \"nj\":\n",
    "        tree = skbio.tree.nj(dm)\n",
    "    else:\n",
    "        raise ValueError(\"Unknown metric: %s. Supported metrics are 'upgma' and 'nj'.\" % metric)\n",
    "\n",
    "    return tree\n",
    "\n",
    "\n",
    "\n",
    "def progressive_msa_and_tree(sequences,\n",
    "                             pairwise_aligner,\n",
    "                             metric=kmer_distance,\n",
    "                             guide_tree=None,\n",
    "                             display_aln=False,\n",
    "                             display_tree=False):\n",
    "    \"\"\" Perform progressive msa of sequences and build a UPGMA tree\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : skbio.SequenceCollection\n",
    "        The sequences to be aligned.\n",
    "    pairwise_aligner : function\n",
    "        Function that should be used to perform the pairwise alignments,\n",
    "        for example skbio.alignment.global_pairwise_align_nucleotide. Must\n",
    "        support skbio.Sequence objects or skbio.TabularMSA objects\n",
    "        as input.\n",
    "    metric : function, optional\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects. This will be used to build a guide tree if one\n",
    "      is not provided.\n",
    "    guide_tree : skbio.TreeNode, optional\n",
    "        The tree that should be used to guide the alignment process.\n",
    "    display_aln : bool, optional\n",
    "        Print the alignment before returning.\n",
    "    display_tree : bool, optional\n",
    "        Print the tree before returning.\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.alignment\n",
    "    skbio.TreeNode\n",
    "    \"\"\"\n",
    "    msa = progressive_msa(sequences, pairwise_aligner=pairwise_aligner,\n",
    "                          guide_tree=guide_tree)\n",
    "\n",
    "    if display_aln:\n",
    "        print(msa)\n",
    "\n",
    "    msa_dm = DistanceMatrix.from_iterable(msa, metric=metric, key='id')\n",
    "    msa_lm = sp.cluster.hierarchy.average(msa_dm.condensed_form())\n",
    "    msa_tree = TreeNode.from_linkage_matrix(msa_lm, msa_dm.ids)\n",
    "    if display_tree:\n",
    "        print(\"\\nOutput tree:\")\n",
    "        d = sp.cluster.hierarchy.dendrogram(msa_lm, labels=msa_dm.ids, orientation='right',\n",
    "                   link_color_func=lambda x: 'black')\n",
    "    return msa, msa_tree\n",
    "\n",
    "def iterative_msa_and_tree(sequences,\n",
    "                           num_iterations,\n",
    "                           pairwise_aligner,\n",
    "                           metric=kmer_distance,\n",
    "                           display_aln=False,\n",
    "                           display_tree=False):\n",
    "    \"\"\" Perform progressive msa of sequences and build a UPGMA tree\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : skbio.SequenceCollection\n",
    "       The sequences to be aligned.\n",
    "    num_iterations : int\n",
    "       The number of iterations of progressive multiple sequence alignment to\n",
    "       perform. Must be greater than zero and less than five.\n",
    "    pairwise_aligner : function\n",
    "       Function that should be used to perform the pairwise alignments,\n",
    "       for example skbio.alignment.global_pairwise_align_nucleotide. Must\n",
    "       support skbio.Sequence objects or skbio.TabularMSA objects\n",
    "       as input.\n",
    "    metric : function, optional\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects. This will be used to build a guide tree if one\n",
    "      is not provided.\n",
    "    display_aln : bool, optional\n",
    "       Print the alignment before returning.\n",
    "    display_tree : bool, optional\n",
    "       Print the tree before returning.\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.alignment\n",
    "    skbio.TreeNode\n",
    "   \"\"\"\n",
    "    if num_iterations > 5:\n",
    "        raise ValueError(\"A maximum of five iterations is allowed.\"\n",
    "                         \"You requested %d.\" % num_iterations)\n",
    "    previous_iter_tree = None\n",
    "    for i in range(num_iterations):\n",
    "        if i == (num_iterations - 1):\n",
    "            # only display the last iteration\n",
    "            display = True\n",
    "        else:\n",
    "            display = False\n",
    "        previous_iter_msa, previous_iter_tree = \\\n",
    "            progressive_msa_and_tree(sequences,\n",
    "             pairwise_aligner=pairwise_aligner,\n",
    "             metric=metric,\n",
    "             guide_tree=previous_iter_tree,\n",
    "             display_aln=display_aln and display,\n",
    "             display_tree=display_tree and display)\n",
    "\n",
    "    return previous_iter_msa, previous_iter_tree\n",
    "\n",
    "def random_sequence(moltype, length):\n",
    "    result = []\n",
    "    alphabet = list(moltype.nondegenerate_chars)\n",
    "    for e in range(length):\n",
    "        result.append(random.choice(alphabet))\n",
    "    return moltype(''.join(result))\n",
    "\n",
    "## Sequence simulation\n",
    "def evolve_sequence(sequence, substitution_probability, indel_probability):\n",
    "    result = []\n",
    "    sequence_length = len(sequence)\n",
    "    insertion_choices = list(sequence.nondegenerate_chars)\n",
    "\n",
    "    result = []\n",
    "    i = 0\n",
    "    while i < sequence_length:\n",
    "        current_char = sequence[i]\n",
    "        if random.random() < substitution_probability:\n",
    "            # simulate a substitution event by adding a character other than the current\n",
    "            # character to the result at this position\n",
    "            substituted_base = random.choice([r for r in sequence.nondegenerate_chars if r != current_char])\n",
    "            result.append(substituted_base)\n",
    "            i += 1\n",
    "        elif random.random() < indel_probability:\n",
    "            # simulate either an insertion or a deletion event. the length of the insertion or\n",
    "            # deletion is determined at random, with shorter lengths being more probable\n",
    "            length = int(np.random.triangular(1, 1, 10))\n",
    "            if np.random.binomial(1, 0.5) == 0:\n",
    "                # simulate an insertion by adding length random characters from\n",
    "                # this sequence's alphabet\n",
    "                result.extend(np.random.choice(insertion_choices, size=length))\n",
    "                i += 1\n",
    "            else:\n",
    "                # simulate a deletion by not appending any of the next length\n",
    "                # characters\n",
    "                i += length\n",
    "        else:\n",
    "            # simulate no mutation occurring\n",
    "            result.append(str(current_char))\n",
    "            i += 1\n",
    "\n",
    "    return sequence.__class__(''.join(result))\n",
    "\n",
    "def evolve_generation(sequence,\n",
    "                      substitution_probability,\n",
    "                      indel_probability,\n",
    "                      increased_rate_probability,\n",
    "                      fold_rate_increase):\n",
    "    child1 = evolve_sequence(sequence, substitution_probability, indel_probability)\n",
    "    if random.random() < increased_rate_probability:\n",
    "        child2 = evolve_sequence(sequence, fold_rate_increase * substitution_probability, indel_probability)\n",
    "    else:\n",
    "        child2 = evolve_sequence(sequence, substitution_probability, indel_probability)\n",
    "    return child1, child2\n",
    "\n",
    "def evolve_generations(ancestral_sequence, generations, substitution_probability,\n",
    "                       indel_probability, increased_rate_probability=0.0,\n",
    "                       fold_rate_increase=5, verbose=False):\n",
    "    \"\"\"\n",
    "    ancestral_sequence : skbio.Sequence object\n",
    "        The sequence that should be used as the root node in the tree.\n",
    "    generations : int\n",
    "        The number of generations to simulate. Must be greater than zero.\n",
    "    substitution_probability : float\n",
    "        The probability at which a substitution mutation should occur.\n",
    "    indel_probability : float\n",
    "        The probability at which either an insertion or a deletion mutation\n",
    "        should occur. One of these will be simulated with even probability,\n",
    "        and the length will be randomly chosen between 1 and 9, with shorter\n",
    "        lengths being more probable.\n",
    "    increased_rate_probability : float, optional\n",
    "        The probability at which child2 should experience an increased rate\n",
    "        of substitution mutations. Default is that this never occurs.\n",
    "    fold_rate_increase : int, optional\n",
    "        If increased_rate_probability is greater than zero, the fold increase\n",
    "        that substitutions should now occur at.\n",
    "    verbose : bool, optional\n",
    "        If True, print the sequences that are simulated at each generation.\n",
    "    \"\"\"\n",
    "    # initial some values and perform some basic error checking\n",
    "    assert generations > 0, \"Must simulate one or more generations.\"\n",
    "\n",
    "    # initialize a list of the previous generations sequences - this gets used\n",
    "    # in the for loop below. since we'll start with the first generation of\n",
    "    # children, root_sequence is the previous generation's sequence\n",
    "    ancestral_sequence.metadata['id'] = '0'\n",
    "    previous_generation_sequences = [ancestral_sequence]\n",
    "\n",
    "    # iterate for each requested generation\n",
    "    for i in range(generations):\n",
    "        # print the generation number and the current number of sequences\n",
    "        if verbose:\n",
    "            print(\"Generation: %d (Number of parent sequences: %d)\" % (i,2**i))\n",
    "            print(\"%s (last common ancestor)\" % ancestral_sequence)\n",
    "            print(\"\")\n",
    "\n",
    "        # create a list to store the current generation of sequences\n",
    "        current_generation_sequences = []\n",
    "\n",
    "        # iterate over the sequences of the previous generation\n",
    "        for parent_sequence in previous_generation_sequences:\n",
    "            # evolve two child sequences - currently the mutation probabilities are\n",
    "            # constant, but should update that to change with generations\n",
    "            r1, r2 = evolve_generation(parent_sequence, substitution_probability, indel_probability,\n",
    "                                       increased_rate_probability, fold_rate_increase)\n",
    "            r1.metadata['id'] = '%s.1' % (parent_sequence.metadata['id'])\n",
    "            r2.metadata['id'] = '%s.2' % (parent_sequence.metadata['id'])\n",
    "            current_generation_sequences.extend([r1, r2])\n",
    "            if verbose:\n",
    "                # if the caller specified verbose output, print the actual sequences\n",
    "                print(\"%s (parent id: %s)\" % (parent_sequence, parent_sequence.metadata['id']))\n",
    "                print(\"%s (child 1 id: %s)\" % (r1, r1.metadata['id']))\n",
    "                print(\"%s (child 2 id: %s)\" % (r2, r2.metadata['id']))\n",
    "                print(\"\")\n",
    "        # current_generation_sequences becomes the next generation's\n",
    "        # previous_generation_sequences\n",
    "        previous_generation_sequences = current_generation_sequences\n",
    "\n",
    "    # upon completion of all generations, return the last generation's sequences\n",
    "    return previous_generation_sequences"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
