{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n",
      "CCTCGGAGGGTAATCTACATTCAAATCATCTCTA\n",
      "CTTGTAAGCCCATTAACAACGACTCATCGCAA\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'show' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-4cfe96ce7294>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    111\u001b[0m                     ids=['a', 'b', 'c'])\n\u001b[1;32m    112\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 113\u001b[0;31m \u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmap\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m'Greens'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;31m# We will use the scikit-bio to create a skbio.distancematrix object. These objects can be viewed as heatmaps.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'show' is not defined"
     ]
    }
   ],
   "source": [
    "# Python Problems 2\n",
    "# Luis J. Ramirez\n",
    "# In this assignment, we work on phylogenitic reconstruction problems that are presented by Dr. Greg Caporaso.\n",
    "\n",
    "# Before we can reconstruct a phylogenetic tree from a DNA sequence, we must first simulate the process of evolution\n",
    "# for a DNA sequence.\n",
    "# We are going to use a python function to model a sequence evolution, then we will run that function to simulate\n",
    "# multiple generations of evolution.\n",
    "\n",
    "import pylab as p\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import random\n",
    "from IPython.core import page\n",
    "page.page = print\n",
    "\n",
    "# First we take a look at the function that is going to help us model this DNA sequence evolution.\n",
    "# This is were most of the modeling takes place.\n",
    "\n",
    "from BioinformaticsCode.algorithms import evolve_sequence\n",
    "\n",
    "# This function will help us model a single generation of a single sequence.\n",
    "# Here, clonal reproduction will occur\n",
    "\n",
    "from BioinformaticsCode.algorithms import evolve_generation\n",
    "\n",
    "# Then we have the entry point function. We provide the parameters for the simulation: Initial sequence, number of\n",
    "# generations and mutation probabilities.\n",
    "# This code is built on the previously mentioned functions.\n",
    "\n",
    "from BioinformaticsCode.algorithms  import evolve_generations\n",
    "\n",
    "# now we can run a simulation using a random DNA sequence and evolve three generations. We will use evolve_sequence from\n",
    "# algorithms and pass the verbose = true parameter.\n",
    "\n",
    "from BioinformaticsCode.algorithms  import random_sequence\n",
    "\n",
    "import skbio\n",
    "\n",
    "sequence = random_sequence(skbio.DNA, 50)\n",
    "\n",
    "sequences = evolve_generations(sequence, 3, 0.1, 0.05, 0.1, True)\n",
    "\n",
    "# sequences will contain the the child sequence from the last generation. We also note that we will only have sequences\n",
    "# from the last generation. Not the ancestral sequence, since this models a real world application.\n",
    "\n",
    "# Lets take a look at the following:\n",
    "\n",
    "print(len(sequences))\n",
    "\n",
    "print(sequences[0])\n",
    "\n",
    "print(sequences[-1])\n",
    "\n",
    "# In our simulation each, each sequence is derived from exactly one sequence from the previous generation, and the\n",
    "# evolution traces back to the ancestral sequence that we provided. This means that are final sequences are all\n",
    "# homologous. Our goal now, is to reconstruct the phylogeny given only the last generation of sequences. We will use the\n",
    "# fact that we know the previous phylogeny to evaluate our performance.\n",
    "\n",
    "# Now we are going to simulate 10 generations of sequences. We will randomly select some of the sequences for the\n",
    "# remainder of this code. We are setting the indel probability to 0.0. This means we wont have to align the selected\n",
    "# This saves us some runtime.\n",
    "\n",
    "from skbio.alignment import global_pairwise_align_nucleotide\n",
    "from BioinformaticsCode.algorithms import progressive_msa\n",
    "from functools import partial\n",
    "\n",
    "indel_probability = 0.0\n",
    "\n",
    "sequences = evolve_generations(sequence, 10, .03, 0.0, 0.1, False)\n",
    "\n",
    "sequences = random.sample(sequences, 25)\n",
    "\n",
    "if indel_probability == 0:\n",
    "    sequences_aligned = sequences\n",
    "else:\n",
    "    gpa = partial(global_pairwise_align_nucleotide, True)\n",
    "    sequences_aligned = progressive_msa(sequences, pairwise_aligner=gpa)\n",
    "\n",
    "\n",
    "# While the simulations are strong for comparing algorithms, they can often be misleading. This is because we simplify ,\n",
    "# the evolutionary process. In practice a combination of real data and simulated data can help/\n",
    "\n",
    "# Now we start to compute phylogenetic trees, we're going to need a way to visualize them. the ete3 python package can\n",
    "# help us do this. We will configure TreeStyle to define the way trees work.\n",
    "\n",
    "import ete3\n",
    "ts = ete3.TreeStyle()\n",
    "ts.show_leaf_name = True\n",
    "ts.scale = 250\n",
    "ts.branch_vertical_margin = 15\n",
    "\n",
    "# We can apply this tree style to a random tree as follows.\n",
    "\n",
    "t = ete3.Tree()\n",
    "t.populate(10)\n",
    "t.render(\"%%inline\", tree_style=ts)\n",
    "\n",
    "# distance based methods to phylogenetic reconstruction\n",
    "\n",
    "# for the next approach, we will rely on computing the distances between the sequences.\n",
    "\n",
    "# We will use dissimilarity distance between two objects x and y. Literature on this can be found online, for now\n",
    "# we are going to show the code.\n",
    "\n",
    "from skbio import DistanceMatrix\n",
    "\n",
    "dm = DistanceMatrix([[0.0, 1.0, 2.0],\n",
    "                     [1.0, 0.0, 3.0],\n",
    "                     [2.0, 3.0, 0.0]],\n",
    "                    ids=['a', 'b', 'c'])\n",
    "\n",
    "_ = dm.plot(cmap ='Greens')\n",
    "\n",
    "# We will use the scikit-bio to create a skbio.distancematrix object. These objects can be viewed as heatmaps. \n",
    "\n",
    "from BioinformaticsCode.algorithms import kmer_distance\n",
    "kmer_dm = DistanceMatrix.from_iterable(sequences, metric=kmer_distance, key='id')\n",
    "_ = kmer_dm.plot(cmap='Greens', title='3mer distances between sequences')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'show' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-87558fe748d2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'show' is not defined"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
